### ISSUE TITLE
Question: General support for generating openapi-extensions from OData V4 specification file

### ISSUE THREAD

**Issue by fisehara** [oasis-tcs/odata-openapi#179]
OpenAPI3 specification allows to use extension properties ( https://swagger.io/docs/specification/openapi-extensions/ ) in the specification file that are not restricted. The underlying rule is to have them starting with a prefix `x-`.  Extensions can be used in the sections:

- info section
- paths section, individual paths and operations
- operation parameters
- responses
- tags
- security schemes

My question is regarding the expressiveness of OData specification file. Could a specification file have custom fields for any target to add custom information that is outside of the OData specification?

If this is possible or intended how could I contribute to add this custom OData field translated to the openAPI extension field

I've seen this implementation detail:
https://github.com/oasis-tcs/odata-openapi/blob/16479a3bd20bc1b37f2c000c61e7bb40111e2b06/tools/V4-CSDL-to-OpenAPI.xsl#L2327-L2335 and was wondering if this could be transferred into a generic approach.

Thanks!

**Comment by ralfhandl** [comment]
Hi Harald,

An [OData Service Definition (CSDL)](https://docs.oasis-open.org/odata/odata-csdl-json/v4.01/odata-csdl-json-v4.01.html) can be extended via "annotations", and all places where this is possible are called out as "MAY contain annotations".

The translation from OData CSDL to OpenAPI does not have a generic mechanism to translate OData annotations to `x-` extension fields in OpenAPI, nor am I planning to add such a mechanism.

Reason is that some OpenAPI consumers will have their rather fixed expectations on which `x-` fields should be there and how they should be filled. It is rather unlikely that an OData annotation can be "reverse engineered" to match these expectations when run through a generic extension translator.

As a poor mans replacement the XSLT-based "old" tool has a parameter `x-tensions` that allows injecting JSON text (well-formed name-value pairs, separated by commas) into the top-level object of an OpenAPI document:
https://github.com/oasis-tcs/odata-openapi/blob/16479a3bd20bc1b37f2c000c61e7bb40111e2b06/tools/V4-CSDL-to-OpenAPI.xsl#L547-L550

There's no such mechanism yet for the "new" JavaScript tool.

What kinds of OpenAPI extensions would you have liked to add?

Thanks in advance
Ralf

**Comment by ralfhandl** [comment]
Hi Harald,

auto-generated or generic UIs on top of web/REST APIs is a very important use case for SAP, and we have a whole web application framework that is driven by annotations in OData APIs:
- [SAP Fiori Elements](https://experience.sap.com/fiori-design-web/smart-templates/)

Fiori Elements leverages a combination of [standard OASIS annotations](https://github.com/oasis-tcs/odata-vocabularies#further-description-of-this-repository) and [SAP-specific annotations](https://github.com/SAP/odata-vocabularies#sap-vocabularies), most prominently from the
- [SAP UI Vocabulary](https://github.com/SAP/odata-vocabularies/blob/main/vocabularies/UI.md).

The use case of "which fields to render in a table" is solved with the [UI.LineItem](https://github.com/SAP/odata-vocabularies/blob/main/vocabularies/UI.md#LineItem) annotation, combined with the [UI.Importance](https://github.com/SAP/odata-vocabularies/blob/main/vocabularies/UI.md#Importance) annotation: table columns are rendered in the order listed in UI.LineItems, and if not all of the columns fit on the screen the least important ones are hidden first.

Does that solve your problem?

Thanks in advance
Ralf

**Comment by fisehara** [comment]
Hi Ralf,

thanks for the links. I'll study them and see how or if we can utilise them for our use case. 
I've been browsing around and found this repo here: https://github.com/SAP-samples/fiori-elements-incident-management

Is this a good start to understand how the OData is consumed or can you recommend some other resources to check for the actual UI implementation?

Thanks a lot.
Harald

**Comment by ralfhandl** [comment]
Hi Harald,

the SAP-samples project you found is an excellent starting point: it is maintained by the same people that develop Fiori Elements, and they know what they do 😄 

Have fun!
Ralf

**Comment by fisehara** [comment]
Hi Ralf,

just a heads uo. I've studied the [SAP UI Vocabulary](https://github.com/SAP/odata-vocabularies/blob/main/vocabularies/UI.md) and it doesn't align with our approach of data model driven UI rendering. Our approach is to not describe how to render the data, but only describe the data model as explicit as possible. For the moment I'm not following this path now.
Basically we found a 'work around' by adding a detail description containing a json string that can be parsed to solve the reference scheme requirement. This detail description is passed to the openapi json through the translator.

Thanks for the guidance and the details. This issue can be closed for now.

Cheers
Harald
